doc/ipmc/IPMC_Data_Plane_HLD.md

```diff
@@ -0,0 +1,472 @@
+
+ attr.id = SAI_ROUTER_INTERFACE_ATTR_V4_MCAST_ENABLE;
+ sai_status_t status = sai_router_intfs_api->set_router_interface_attribute(port.m_rif_id, &attr);
+```
+
+Example of creating an ipmc group member:
+
+```c++
+sai_attribute_t ipmc_entry_attr;
+vector<sai_attribute_t> ipmc_entry_attrs;
+
+ipmc_entry_attr.id = SAI_IPMC_GROUP_MEMBER_ATTR_IPMC_GROUP_ID;
+ipmc_entry_attr.value.oid = group_id;
+ipmc_entry_attrs.push_back(ipmc_entry_attr);
+
+ipmc_entry_attr.id = SAI_IPMC_GROUP_MEMBER_ATTR_IPMC_OUTPUT_ID;
+ipmc_entry_attr.value.oid = m_rif;
+ipmc_entry_attrs.push_back(ipmc_entry_attr);
+sai_status_t status = sai_ipmc_group_api->create_ipmc_group_member(&m_id, gSwitchId, (uint32_t)ipmc_entry_attrs.size(), ipmc_entry_attrs.data());
+```
+
+Example of creating an ipmc route:
+
+```c++
+sai_attribute_t ipmc_entry_attr;
+vector<sai_attribute_t> ipmc_entry_attrs;
+sai_ipmc_entry_t ipmc_entry;
+ipmc_entry_attr.id = SAI_IPMC_ENTRY_ATTR_OUTPUT_GROUP_ID;
+ipmc_entry_attr.value.oid = ipmc_group_id;
+ipmc_entry_attrs.push_back(ipmc_entry_attr);
+
+ipmc_entry_attr.id = SAI_IPMC_ENTRY_ATTR_PACKET_ACTION;
+ipmc_entry_attr.value.oid = SAI_PACKET_ACTION_FORWARD;
+ipmc_entry_attrs.push_back(ipmc_entry_attr);
+
+ipmc_entry_attr.id = SAI_IPMC_ENTRY_ATTR_RPF_GROUP_ID;
+ipmc_entry_attr.value.oid = rpf_group_id;
+ipmc_entry_attrs.push_back(ipmc_entry_attr);
+
+ipmc_entry.vr_id = vrf_id;
+ipmc_entry.switch_id = gSwitchId;
+ipmc_entry.type = src_ip.isZero() ? SAI_IPMC_ENTRY_TYPE_XG : SAI_IPMC_ENTRY_TYPE_SG;
+copy(ipmc_entry.source, src_ip);
+ipmc_entry.source.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+copy(ipmc_entry.destination, dest_ip);
+ipmc_entry.destination.addr_family = SAI_IP_ADDR_FAMILY_IPV4;
+
+sai_status_t status = sai_ipmc_api->create_ipmc_entry(&ipmc_entry, (uint32_t)ipmc_entry_attrs.size(), ipmc_entry_attrs.data());
+```
+
+### Configuration and management
+
+#### Manifest
+
+N/A
+
+#### CLI/YANG model Enhancements
+
+The IPMC forwarding function is enabled on the interface
+
+##### Interface Yang Model
+
+Revised sonic-interface.yang to add ipmc_forward field
+
+```yang
+container INTERFACE {
+..
+  leaf ipmc_forward {
+      description "enable/disable ipmc forward for the interface";
+      type enumeration {
+          enum enable;
+          enum disable;
+      }
+  }
+..
+}
+```
+
+##### IPMC commands
+
+SONIC Click-based configuration and monitoring CLIs have been introduced in SONIC for interface ipmc
+
+###### Config commands
+
+```python
+config interface ipmc <interface_name> <ipmc_forward>
+```
+
+- Run this command to enable or disable IP multicast packet forwarding on an interface
+- interface_name: interface name (Ethernet/Vlan/PortChannel/sub-interface)
+- ipmc_forward: enable the IP multicast packet forwarding function
+
+###### Show commands
+
+```python
+show interface ipmc { <interface_name> }
+```
+
+- Run the IP multicast packet forwarding command to show the configuration of the multicast packet forwarding function on the interface
+

```

Comment:
It would be helpful to have a CLI to display the IPMC entries and  SAI_IPMC_ENTRY_ATTR_COUNTER_ID